File: algorithms/graph/bellman_ford.py
"""
=====================================================================
Bellman-Ford / Беллман-Форд
---------------------------------------------------------------------
Работает с ОТРИЦАТЕЛЬНЫМИ весами. Детектирует отрицательные циклы.
Time:  O(V·E)
Space: O(V)
=====================================================================
"""
from math import inf
from typing import List, Tuple


def bellman_ford(nodes: List, edges: List[Tuple], start) -> dict:
    dist = {n: inf for n in nodes}
    parent = {}
    dist[start] = 0

    for _ in range(len(nodes) - 1):
        updated = False
        for u, v, w in edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                updated = True
        if not updated:
            break

    # Проверка на отрицательный цикл
    for u, v, w in edges:
        if dist[u] + w < dist[v]:
            raise ValueError("Negative-weight cycle detected")

    return {"dist": dist, "parent": parent}


if __name__ == "__main__":
    nodes = ["A", "B", "C", "D"]
    edges = [
        ("A", "B", 1),
        ("B", "C", -2),
        ("A", "C", 4),
        ("C", "D", 3),
    ]
    print(bellman_ford(nodes, edges, "A")["dist"])
    # {'A': 0, 'B': 1, 'C': -1, 'D': 2}

File: algorithms/graph/bellman_ford.js
// =====================================================================
// Bellman-Ford / Алгоритм Беллмана-Форда
// ---------------------------------------------------------------------
// Кратчайшие пути от стартовой вершины. В отличие от Дейкстры —
// РАБОТАЕТ С ОТРИЦАТЕЛЬНЫМИ ВЕСАМИ и обнаруживает отриц. циклы.
//
// Идея: V-1 раз релаксируем ВСЕ рёбра. На V-й итерации если что-то
//       релаксировалось — есть отрицательный цикл.
// Time:  O(V·E)
// Space: O(V)
//
// Ввод: список рёбер edges = [[u, v, weight], ...], список вершин nodes
// =====================================================================

function bellmanFord(nodes, edges, start) {
    const dist = new Map();
    const parent = new Map();
    for (const n of nodes) dist.set(n, Infinity);
    dist.set(start, 0);

    // V-1 итерация релаксации
    for (let i = 0; i < nodes.length - 1; i++) {
        let updated = false;
        for (const [u, v, w] of edges) {
            if (dist.get(u) + w < dist.get(v)) {
                dist.set(v, dist.get(u) + w);
                parent.set(v, u);
                updated = true;
            }
        }
        if (!updated) break; // early exit — пути стабилизировались
    }

    // V-я итерация: если что-то ещё релаксируется → отрицательный цикл
    for (const [u, v, w] of edges) {
        if (dist.get(u) + w < dist.get(v)) {
            throw new Error('Graph contains negative-weight cycle');
        }
    }

    return { dist, parent };
}

const nodes = ['A', 'B', 'C', 'D'];
const edges = [
    ['A', 'B', 1],
    ['B', 'C', -2],
    ['A', 'C', 4],
    ['C', 'D', 3],
];
const { dist } = bellmanFord(nodes, edges, 'A');
console.log([...dist.entries()]); // A:0, B:1, C:-1, D:2

module.exports = bellmanFord;
